### PR TITLE
S4 t7  create inputting of recurring schedule

### PR DESCRIPTION
**Pull request by Kyle199910**
Description:

https://trello.com/c/Xq8NCoxA/77-s4-t7-create-inputting-of-recurring-schedule
-displays checkboxes for worker to input schedule for days of the week; submit button should make recurring scheduled based on what days are selected

- Added Worker Input Time page link on dashboard

What to test for:

- [x] Once you run npm start or npm run start:dev, make sure your link is "http://localhost:3000/". It will shortly then change it to "http://localhost:3000/login".
- [x]  Once you make it onto the Dashboard, click on Links and check if they are not broken
- [x] Go to the work input time page, enter in a date with the given date input and check any day, check to see if multiple days can be checked and if the done button doesn't break the page
- [x] -After that, right click page, click inspect and go to console to checkout: output should be an array of dates; test multiple dates like checking off Tuesday when the date selected is a Wednesday, check to see if console output Tuesday's starting the next week
- [x] - Also test cases where the user doesn't select a date or check a checkbox, an alert should pop up
Capture

### CODE REVIEW

File: react-ui/package.json [hunk 1]
```diff
@@ -20,5 +20,17 @@
     "eject": "react-scripts eject"
   },
   "proxy": "http://localhost:4000",
-  "devDependencies": {}
+  "devDependencies": {},
+  "browserslist": {
+    "production": [
```
Comment (catai9) [inline react-ui/package.json]: Hi Kyle,
Was this added on purpose or just a side effect of running the code? If it was purposeful, just a side note that we will need to check that our production website still works after this PR is merged.

File: react-ui/src/App.js [hunk 2]
```diff
@@ -9,6 +9,9 @@ import {
   Route,
   Redirect
 } from "react-router-dom";
+import WorkerInputpage from './components/Layouts/InputWorkerTimes/WorkerInputpage'
+
```
Comment (catai9) [inline react-ui/src/App.js]: Hi Kyle,
Minor concern but is it possible to remove the unnecessary whitespaces on line 13, 14 and 15 so we can reduce the file sizes and increase file clarity?

File: react-ui/src/App.js [hunk 3]
```diff
@@ -44,6 +47,10 @@ class App extends Component {
           {/* <Redirect from="/" to="dashboard/" /> */}
         </Route>
 
+        <Route path="/WorkerAvailableInputPage" component={WorkerInputpage}>
```
Comment (catai9) [inline react-ui/src/App.js]: Hi Kyle,
Minor suggestion, but can the path be lowercase instead of uppercase to stay consistent with the rest of the routes, e.g. `inputAvailability` or something along that line?

File: react-ui/src/components/Layouts/Dashboard.js [hunk 4]
```diff
@@ -62,6 +63,12 @@ class Dashboard extends Component {
           <li>
             <Link to={`/Dashboard/CreateAppointment/${this.state.email}`}>Create Appointment</Link>
           </li>
+
+          <li>
+            <Link to ={`/WorkerAvailableInputPage`}>Worker Input Time Page</Link>
+
```
Comment (catai9) [inline react-ui/src/components/Layouts/Dashboard.js]: Hi Kyle,
Minor concern but please remove the unneeded whitespace on line 69.

File: react-ui/src/components/Layouts/Dashboard.js [hunk 5]
```diff
@@ -62,6 +63,12 @@ class Dashboard extends Component {
           <li>
             <Link to={`/Dashboard/CreateAppointment/${this.state.email}`}>Create Appointment</Link>
           </li>
+
+          <li>
+            <Link to ={`/WorkerAvailableInputPage`}>Worker Input Time Page</Link>
```
Comment (catai9) [inline react-ui/src/components/Layouts/Dashboard.js]: Hi Kyle,
I think we might need to also keep it with the Dashboard prefix so that these pages have consistent navigation bars and overall layout, i.e., `/Dashboard/InputAvailability`. Also, I would rename the route to have a verb to increase clarity of the purpose of that specific page.

File: react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css [hunk 6]
```diff
@@ -0,0 +1,23 @@
+.appHeader {
+    height: 40px;
+    background-color: rgb(93, 93, 226);
+}
+
+.HeaderText1 {
+
```
Comment (catai9) [inline react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css]: Hi Kyle,
Please remove unnecessary whitespace to increase code readability.

File: react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css [hunk 7]
```diff
@@ -0,0 +1,23 @@
+.appHeader {
+    height: 40px;
+    background-color: rgb(93, 93, 226);
```
Comment (catai9) [inline react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css]: Hi Kyle,
Side note, but it may be better to leverage the constants in the sass file (yet to be merged) so that our whole application has a similar colour styling.

File: react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+          <br/>
+          <input type="submit" value="Done" />
+        </form>
+         
+      </div>
+    );
+  }
+}
+   export default Check
```
Comment (catai9) [inline react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js]: Hi Kyle,
Minor suggestion, right click and use Format on Visual Studio so that all of your code lines up.

File: react-ui/src/components/Layouts/InputWorkerTimes/WorkerInputpage.css [hunk 9]
```diff
@@ -0,0 +1,30 @@
+
+/* .SelectSchools{
```
Comment (catai9) [inline react-ui/src/components/Layouts/InputWorkerTimes/WorkerInputpage.css]: Hi Kyle,
When code is commented out, it's usually good to leave a description of why it's still kept (its purpose) OR to just remove it to prevent unnecessary growth in file sizes.

File: react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        var item = event.target.value;
+         
+        this.setState(prevState => ({ checkedTimes: prevState.checkedTimes.set(item, isChecked) }));
+  }
+     
+  handleSubmit(event) {
+    console.log("this button works"); // edit this for future functions
+    event.preventDefault();
+  }
+
```
Comment (catai9) [inline react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js]: Hi Kyle,
Same comment as above, just make sure that unneeded whitespace is removed from the files so that they do not become unnecessarily long in length.

File: react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+/* import {Container, Row, Col} from 'react-grid-system'; */
+
+class TimeSlot extends React.Component { // this is a class component for the checkbox feature of selecting recurring days
+  constructor() {
+    super();
+    this.state = {
+      Times: [
+        {id: 1, value: "9:00", endtime:"9:30"},
```
Comment (catai9) [inline react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js]: Hi Kyle,
I would suggest moving this into a constants file in case other functions need it in the future (so unnecessary repetition is avoided). I assume that this is just a temporary measure until the endpoint is made to fetch all the timeslots so another alternative is to write a brief comment saying that this is temporary and should be replaced with that endpoint once completed.

File: react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+     
+  handleChange(event) {
+        var isChecked = event.target.checked;
+        var item = event.target.value;
+         
+        this.setState(prevState => ({ checkedDays: prevState.checkedDays.set(item, isChecked) }));
+  }
+     
+  handleSubmit(event) { // submits selected days and assign recurrenting dates based off boxes checked
+    /* console.log(this.state);  */ //test line
```
Comment (catai9) [inline react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js]: Hi Kyle,
Not sure if these test lines should be kept? Will they be used in the future or can they be removed entirely?

**Review by catai9 — CHANGES_REQUESTED** [review]
Hi Kyle,
Overall, really great job! I ran it locally and it looks super nice! There's a few minor changes with removing whitespace from a few files and formatting the code in the files (right click and use Format on Visual Studio). Additionally, I would suggest writing a unit test for this component (after PR #31 gets merged). I love the error checking for the user inputs!

File: react-ui/package.json (hunk 1 above)
Comment (Kyle199910) [reply react-ui/package.json]: that was a side effect of running the code

File: react-ui/src/App.js (hunk 2 above)
Comment (Kyle199910) [reply react-ui/src/App.js]: Deleted Whitespace as requested

File: react-ui/src/App.js (hunk 3 above)
Comment (Kyle199910) [reply react-ui/src/App.js]: Adjusted the naming as requested

File: react-ui/src/components/Layouts/Dashboard.js (hunk 5 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/Dashboard.js]: renamed as requested, currently not sure on how to adjust the routing, I had the link redirect the user to another page

File: react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css (hunk 7 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css]: If I set up my code routing to work like the other links on the page, then the appheader file can be deleted entirely

File: react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js (hunk 12 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js]: i was just using them for debugging purposes, they can be removed entirely

File: react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js (hunk 11 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js]: assuming that Melissa has worked on S4-T6, this code will not be needed, but for now, i put a comment stating this is a temporary measure

File: react-ui/src/components/Layouts/InputWorkerTimes/WorkerInputpage.css (hunk 9 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/WorkerInputpage.css]: Noted for future iterations of coding

File: react-ui/package.json (hunk 1 above)
Comment (catai9) [reply react-ui/package.json]: Hi Kyle,
Makes sense. Thank you for removing it!

File: react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css (hunk 7 above)
Comment (catai9) [reply react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css]: Hi Kyle, 
Yes, you can remove it if it's not needed so the website has a consistent look and feel. If you do have a specific reason as to why you want to keep it, just make a comment and you can also leave it as well.

File: react-ui/src/components/Layouts/Dashboard.js (hunk 5 above)
Comment (catai9) [reply react-ui/src/components/Layouts/Dashboard.js]: Hi Kyle,
To adjust the routing, add in the relevant **Link** into the Dashboard component (similar to how it is currently done for the `CreateAppointment` route. The difference would be that your route should show when the UserType is `worker` rather than `student`. In addition, you would also need to add a **Route** at the bottom of that file to go to your InputAvailability component.

File: react-ui/src/App.js [hunk 13]
```diff
@@ -44,6 +45,10 @@ class App extends Component {
           {/* <Redirect from="/" to="dashboard/" /> */}
         </Route>
 
+        <Route path="/workerinputavailability" component={WorkerInputpage}>
```
Comment (catai9) [inline react-ui/src/App.js]: Hi Kyle,
I would remove this and add it to the `Dashboard` component rather than the App section since it should only be available after the user has logged into the system.

File: react-ui/src/components/Layouts/Dashboard.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
+import React, { Component, Fragment } from "react";
```
Comment (catai9) [inline react-ui/src/components/Layouts/Dashboard.js]: Hi Kyle,
This file was removed in a past PR. When updating your branch with master, make sure to resolve conflicts appropriately. My tip would be to look at the files you changed on your PR to make sure there was no undesirable changes in your PR.

File: react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css [hunk 15]
```diff
@@ -0,0 +1,23 @@
+.appHeader {
```
Comment (catai9) [inline react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css]: Hi Kyle,
Make sure to change the file extension to `scss` instead of `css` for consistency.

**Review by catai9 — CHANGES_REQUESTED** [review]
Hi Kyle,

Thank you for making the desired improvements. There are a few more tweaks that I would recommend. A helpful tip would be to use the Files Changed tab on your GitHub PR to see all the files and code that you have changed. If you noticed anything out of the ordinary or excessive whitespace, please address them in your PR (it may be side effects of resolving any merge conflicts). Additionally, please write brief unit tests for this component before we merge it into master. Great work so far!

File: react-ui/src/components/Layouts/Dashboard.js (hunk 5 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/Dashboard.js]: Integrated the pathing with the new dashboard provided.

File: react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css (hunk 7 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css]: Code has been removed

File: react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js (hunk 11 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js]: code has been deleted for now, if needed, I can retrieve it from a previous branch

File: react-ui/src/App.js (hunk 13 above)
Comment (Kyle199910) [reply react-ui/src/App.js]: done

File: react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css (hunk 15 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/AvailableMeetingHeader.css]: all css files have been changed to scss

File: react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js (hunk 10 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js]: > Hi Kyle,
> 
> Thank you for making the desired improvements. There are a few more tweaks that I would recommend. A helpful tip would be to use the Files Changed tab on your GitHub PR to see all the files and code that you have changed. If you noticed anything out of the ordinary or excessive whitespace, please address them in your PR (it may be side effects of resolving any merge conflicts). Additionally, please write brief unit tests for this component before we merge it into master. Great work so far!

Hi Amy, how would I exactly write a unit test?

Comment (catai9) [reply react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js]: Hi Kyle,
I would suggest looking at the unit tests for the AppointmentList component located in the respective folder under src. Unit test files end in `.test.js`.

Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/TimeSlot.js]: Hi Amy, 
I have written some unit tests as requested.

**Comment by catai9** [comment]
Hi Kyle,

As discussed, I understand that the styling of this section and the integration of this feature with the back end are in scope for another ticket. 

I have tested the code in the following manner:
1. Ran the code locally and ensured that the sign up form renders and the popup shows after clicking the "Done" button. Also tried not selecting a day of the week and made sure an error popup showed. Also tried selecting a date from the Calendar and ensured that an error message showed. 
image
2. Ran the automated unit tests in the `react-ui` folder. All tests passed, but I noticed that the `CheckboxApplication.js` had only 48.57% file coverage. Please review and update the tests to ensure all code lines are covered. The Uncovered Line section will be a good indicator of which lines to focus on (e.g., line 62, 65, 66, etc.).

File: server/package.json [hunk 16]
```diff
@@ -12,7 +12,8 @@
     "express": "^4.16.2",
     "joi": "^17.2.1",
     "morgan": "^1.9.0",
-    "pg": "^7.3.0",
+    "native": "^0.3.3",
```
Comment (catai9) [inline server/package.json]: Hi Kyle, 
Not sure if this dependency is used anywhere. To prevent unnecessarily large dependencies, I would recommend removing this dependency if it is not used.

File: react-ui/src/components/Layouts/InputWorkerTimes/WorkerInputpage.js [hunk 17]
```diff
@@ -0,0 +1,39 @@
+import React from 'react'
+import Check from './CheckboxApplication';
+import './WorkerInputpage.scss';
+import { Link } from 'react-router-dom';
+import { Container, Row, Col } from 'react-grid-system';
+import { Redirect, Route, withRouter } from "react-router-dom";
+import ReactDOM from "react-dom";
+import moment from 'moment';
+
```
Comment (catai9) [inline react-ui/src/components/Layouts/InputWorkerTimes/WorkerInputpage.js:9]: Hi Kyle,
If possible, I would recommend removing the whitespaces on lines 10 to 13 to reduce unnecessarily long files.

File: .gitignore [hunk 18]
```diff
@@ -23,4 +23,6 @@ server/.jest
 # misc
 react-ui/npm-debug.log*
 react-ui/yarn-debug.log*
-react-ui/yarn-error.log*
\ No newline at end of file
+react-ui/yarn-error.log*
+react-ui/package.json
+react-ui/package.json
```
Comment (catai9) [inline .gitignore]: Hi Kyle,
We want to keep the package.json in GitHub so Heroku can view and install necessary dependencies for the front end. I would recommend removing lines 27 and 28 unless they are needed for another reason.

File: react-ui/package.json [hunk 19]
```diff
@@ -28,5 +28,11 @@
   "devDependencies": {
     "@babel/plugin-proposal-decorators": "^7.12.1",
     "jest-environment-jsdom-sixteen": "^1.0.3"
-  }
+  },
```
Comment (catai9) [inline react-ui/package.json:31]: Hi Kyle,
As mentioned before (this may be a side effect of running the code), but please remove the browserslist to prevent deployment issues.

File: react-ui/src/App.js [hunk 20]
```diff
@@ -2,6 +2,7 @@ import React, { Component } from 'react'
 import LogIn from "./components/Layouts/LogIn"
 import Dashboard from "./components/Dashboard/Dashboard"
 import Signup from "./components/Layouts/Signup"
+import WorkerInputpage from './components/Layouts/InputWorkerTimes/WorkerInputpage'
```
Comment (catai9) [inline react-ui/src/App.js]: Hi Kyle,
This import statement is no longer needed. I would recommend removing it to reduce future confusion.

**Review by catai9 — CHANGES_REQUESTED** [review]
Hi Kyle,
Great work so far. Thank you for making the requested changes. There are some other comments I would like your opinions on and I would suggest removing the `Layouts/Dashboard.js` file as it has been already moved to another folder (it may have been a side effect of a merge conflict resolution).

File: react-ui/src/components/Layouts/InputWorkerTimes/WorkerInputpage.js (hunk 17 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/WorkerInputpage.js:9]: Done

File: react-ui/package.json (hunk 19 above)
Comment (Kyle199910) [reply react-ui/package.json:31]: Removed

File: react-ui/src/App.js (hunk 20 above)
Comment (Kyle199910) [reply react-ui/src/App.js]: Removed import statement

File: server/package.json (hunk 16 above)
Comment (Kyle199910) [reply server/package.json]: removed native dependency

File: .gitignore (hunk 18 above)
Comment (Kyle199910) [reply .gitignore]: alright I have removed lines 27 and lines 28

**Comment by Kyle199910** [comment]
> Hi Kyle,
> 
> As discussed, I understand that the styling of this section and the integration of this feature with the back end are in scope for another ticket.
> 
> I have tested the code in the following manner:
> 
> 1. Ran the code locally and ensured that the sign up form renders and the popup shows after clicking the "Done" button. Also tried not selecting a day of the week and made sure an error popup showed. Also tried selecting a date from the Calendar and ensured that an error message showed.
>    image
> 2. Ran the automated unit tests in the `react-ui` folder. All tests passed, but I noticed that the `CheckboxApplication.js` had only 48.57% file coverage. Please review and update the tests to ensure all code lines are covered. The Uncovered Line section will be a good indicator of which lines to focus on (e.g., line 62, 65, 66, etc.).

Hi Amy, I have added more units to ensure that the NPM Test for the checkbox application had more coverage, now results should be at 100%.

File: react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+    this.handleChange = this.handleChange.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+  }
+
+  handleChange(event) {
+    var isChecked = event.target.checked;
+    var item = event.target.value;
+    this.setState(prevState => ({ checkedDays: prevState.checkedDays.set(item, isChecked) }));
+  }
+
```
Comment (bradleyhallett) [inline react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js:32]: I might add a small explanation or "instructions" as to what the date means as I didn't understand at first why I need to to select days of the week AND a date on the calendar. Only after using the application a couple times and seeing the output in the console did I understand what the select date meant

**Review by bradleyhallett — CHANGES_REQUESTED** [review]
Possibly add some instruction as to why to user is selecting a date, and how many weeks the schedule will be repeated

File: react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js (hunk 21 above)
Comment (Kyle199910) [reply react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js:32]: Hi Bradley,

I have added a header outlining the basic instructions.
Capture


Hopefully, that eliminates the lack of insight for a first time user.

Comment (bradleyhallett) [reply react-ui/src/components/Layouts/InputWorkerTimes/CheckboxApplication.js:32]: Perfect! Thank you Kyle, It looks great!

**Review by bradleyhallett — APPROVED** [review]

**Review by catai9 — APPROVED** [review]
HI Kyle,
Thank you for updating the tests. 
1. I now have 100% file coverage for the CheckboxApplication component (and 100% for AppointmentList component) as shown below when I run the automated front end tests. As previously discussed, I understand that the testing for the WorkerInputpage component will be in scope for another task (and out of scope for this PR) and the low coverage for the other front end unit tests are in scope for another task (and out of scope for this PR). 
image.
2. I also ran the automated server tests and ensured that there is 100% file coverage and all tests passed (with the exception of timeslot.js  that has lower coverage than 100%, but is currently in scope for another PR).
3. I also ran the code locally and obtained the following screen when I logged in with worker credentials. I also tried various combinations of valid and invalid inputs and ensured that the result was as expected (i.e., the appropriate error popups appeared for invalid inputs and the appropriate successful popup appeared for valid inputs). As mentioned before, I understand that the styling of this component along with the integration to the back end are in scope for another ticket (and subsequently out of scope for this PR). 
image

As such, I am approving my side of the PR review and bequeath to Bradley the responsibility of further reviewing specific code in the PR and merging in the code (once he is satisfied).
